### ISSUE TITLE
Feature request: Option to use lowercase hex literals

### ISSUE THREAD

**Issue by DavidBuchanan314** [astral-sh/ruff#10432]
For a brief time window, `black` had this feature, before it was reverted https://github.com/psf/black/issues/1692

I'm a big fan of lowercase hex literals, for most of the practical reasons outlined in the above issue, and also for purely aesthetic reasons - uppercase feels shouty.

If I had my way I'd make lowercase the default, but failing that it'd be great to have a configuration option.

**Comment by MichaReiser** [comment]
Hi @DavidBuchanan314 

If I understand correctly, this is about formatting hex number literals like `0xEF`? I'm asking because Ruff normalizes hex numbers in string literals to lowercase [Playground](https://play.ruff.rs/57e30c6a-4d5e-4227-818a-d0c9dbe25d87)

My preference would be to change the default to lowercase to make it consistent with hex formatting in strings. However, from reading through Black's issues and discussions on the string hex formatting PRs, I understand that Black changed the preferred casing for hex number literals multiple times and that it's unlikely that they'll change it again in the future. I don't know if I want to deviate from Black's default, considering that the benefits are marginal. 

That means that supporting lowercase hex formatting most likely requires adding a new formatter option. I'm not opposed to this, but we need to make a holistic decision if we want to support formatting-related options or keep Black's opinionated stance to have no-formatting-related settings (or very limited).

**Comment by DavidBuchanan314** [comment]
Yeah, it's the number literals I'm interested in. As for whether it's actually worth it, I guess I'll leave that to everyone else to decide :)

**Comment by kaddkaka** [comment]
To me this kind of configuration is great. It shouldn't affect other formatting and it allows to adopt ruff formatting on an existing codebase that uses either style.

I would also like to propose three values for this kind of config:

- uppercase
- lowercase
- dont_touch (any_case) 

The last option let's the programmer choose case and don't modify it with the formatter - this should make the formatter faster 🐎

**Comment by DavidBuchanan314** [comment]
I was just thinking about this again, and another argument for lower-case is that the `hex()` built-in produces lower-case hex.

```py
>>> hex(123)
'0x7b'
```

**Comment by PhilipYip1988** [comment]
I was looking for this option too. The formal representation for hexadecimal values in the ipython console is in lower case:

```python
>>> 0xABB4AB8A
2880744330

>>> hex(2880744330)
'0xabb4ab8a'

>>> bin(200)
'0b11001000'
```

At a quick glance, the B and 8 can be mistaken and the 4 and A can also be mistaken. The characters are much more distinct in lower case...

For the exponent lower case e is preferred but black is consistent here:

```python
>>>1E100
1e+100
```
The formal representation of strings also prefers single-quotations, unless a string literal is included:

```python
>>> "hello world!"
'hello world!'

>>>'text = \'hello world!\''
"text = 'hello world!'"
```

The Ruff option quote-style fixes this:

```python
[format]
# Prefer single quotes over double quotes.
quote-style = "single"
```

Lower case f is preferred for a formatted string and black is consistent here:

```python
var = 'world'
f'hello {var}!'
```

The prefix for raw strings are not changed in black because code syntax highlighters differentiate R and r for file paths and regular expressions: 

```python
file_path = R'C:\Windows\System32'
email = r'[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}'
```

Having something similar to the following would be useful:

```python
[format]
# Prefer single quotes over double quotes.
quote-style = 'single'
# Prefer lowercase in hexadecimal and binary.
hex-style = 'lower'
```

If there are other changes that black makes, that differ to the formal representation and the style seen in the official Python Documentation itself, then it might also be convenient for end users if these changes are grouped. For example:

```python
# Match Python's official documentation code style
style='python'
```

```python
# Match black code style
style='black'
```

**Comment by hodunov** [comment]
Any updates on this? It really makes sense.

**Comment by avabarnard** [comment]
Commenting because @MichaReiser said there needs to be more demand to add this so hopefully you reconsider at some point soon. Came here looking for an option to do exactly what kaddkaka suggested. I'm trying to introduce ruff to an existing codebase and its reformatting the hex in an ugly way. Also, it seems like a strange choice to have the inconsistency in formatting hex numbers inside and outside of string literals but I'm not here to argue with tradition, just want the option.

**Comment by laura240406** [comment]
I'm also trying to switch to ruff but the uppercase hex is a huge issue for me.

**Comment by inno** [comment]
@MichaReiser What would constitute "enough demand" in order to get https://github.com/astral-sh/ruff/issues/10432 reopened and merged?

**Comment by MichaReiser** [comment]
We don't have any absolute number, but it has to be something with high demand in the community: a feature that is requested frequently by many users.

**Comment by laura240406** [comment]
What are the downsides of including this? There are multiple people who would benefit from it while I don't see any reason why this would harm the user experience of people who don't need it.

**Comment by GioeleBucci** [comment]
Any updates on this? It would be really neat to leave the hex literals as the programmer has written them.
